src/main/java/org/folio/circulation/services/agedtolost/AssignLostFeesWhenAgedToLostService.java

```diff
@@ -0,0 +1,286 @@
+
+        return succeeded(LoanToAssignFees.usingLoans(allLoans))
+          .after(this::fetchFeeFineOwners)
+          .thenComposeAsync(this::fetchFeeFineTypes)
+          .thenCompose(this::assignLostFeesForLoans);
+      }));
+  }
+
+  private CompletableFuture<Result<Void>> assignLostFeesForLoans(
+    Result<List<LoanToAssignFees>> loansToAssignFeesResult) {
+
+    return loansToAssignFeesResult
+      .after(loansToAssignFees -> allOf(loansToAssignFees, this::assignLostFeesForLoan))
+      .thenApply(r -> r.map(notUsed -> null));
+  }
+
+  private CompletableFuture<Result<Loan>> assignLostFeesForLoan(LoanToAssignFees loanToAssignFees) {
+    return createAccountsForLoan(loanToAssignFees)
+      .after(feeFineFacade::createAccounts)
+      .thenCompose(r -> r.after(notUsed -> updateLoanBillingInfo(loanToAssignFees)));
+  }
+
+  private CompletableFuture<Result<Loan>> updateLoanBillingInfo(LoanToAssignFees loanToAssignFees) {
+    final Loan updatedLoan = loanToAssignFees.getLoan().setLostItemHasBeenBilled();
+
+    return loanRepository.updateLoan(updatedLoan);
+  }
+
+  private Result<List<CreateAccountCommand>> createAccountsForLoan(LoanToAssignFees loanToAssignFees) {
+    return validateCanCreateAccountForLoan(loanToAssignFees)
+      .map(notUsed -> {
+        final LostItemPolicy lostItemPolicy = loanToAssignFees.getLostItemPolicy();
+
+        return Stream.of(pair(lostItemPolicy.getSetCostFee(), loanToAssignFees.getLostItemFeeType()),
+            pair(lostItemPolicy.getProcessingFee(), loanToAssignFees.getLostItemProcessingFeeType()))
+          .filter(pair -> pair.getKey().isChargeable())
+          .map(pair -> CreateAccountCommand.builder()
+            .withAmount(pair.getKey().getAmount())
+            .withCreatedByAutomatedProcess(true)
+            .withFeeFine(pair.getValue())
+            .withFeeFineOwner(loanToAssignFees.getOwner())
+            .withLoan(loanToAssignFees.getLoan())
+            .withItem(loanToAssignFees.getLoan().getItem())
+            .build())
+          .collect(Collectors.toList());
+      });
+  }
+
+  private CompletableFuture<Result<List<LoanToAssignFees>>> fetchFeeFineTypes(
+    Result<List<LoanToAssignFees>> allLoansResult) {
+
+    return feeFineRepository.getAutomaticFeeFines(lostItemFeeTypes())
+      .thenApply(r -> r.combine(allLoansResult, this::mapFeeFineTypesToLoans));
+  }
+
+  private List<LoanToAssignFees> mapFeeFineTypesToLoans(Collection<FeeFine> feeTypes,
+     List<LoanToAssignFees> allLoans) {
+
+    return allLoans.stream()
+      .map(loanToAssignFees -> loanToAssignFees.withFeeFineTypes(feeTypes))
+      .collect(Collectors.toList());
+  }
+
+  private CompletableFuture<Result<List<LoanToAssignFees>>> fetchFeeFineOwners(
+    List<LoanToAssignFees> allLoans) {
+
+    final Set<String> uniqueEffectiveLocationServicePoints = allLoans.stream()
+      .map(LoanToAssignFees::getOwnerServicePointId)
+      .collect(Collectors.toSet());
```

Comment:
I think this kind of logic is likely repeated in other places, could this be extracted out into a higher level function e.g.

```
    val uniqueEffectiveLocationServicePoints
      = uniqueSetOf(allLoans, LoanToAssignFees::getOwnerServicePointId);
```

```
  private <T, R> Set<R> uniqueSetOf(List<T> collection, Function<T, R> mapper) {
    return collection.stream()
      .map(mapper)
      .collect(Collectors.toSet());
  }
```

What do you think?

Reply:
Sounds good, I've applied this suggestion.